autofit/graphical/expectation_propagation/ep_mean_field.py

```diff
@@ -223,3 +326,165 @@ def __repr__(self) -> str:
+
+    factor_approximation(factor)
+        create the FactorApproximation for the factor
+
+    project_factor_approx(factor_approximation)
+        given the passed FactorApproximation, return a new `EPMeanField`
+        object encoding the updated mean-field approximation
+    """
+
+    def __init__(
+        self,
+        factor_graph: FactorGraph,
+        factor_mean_field: Dict[Factor, MeanField],
+        factor_rescale: Dict[Factor, Dict[Variable, float]],
+        factor_subset_factor: Dict[Factor, Factor],
+        ep_mean_field: EPMeanField,
+        plates_index: Dict[Plate, List[int]],
+    ):
+        super().__init__(factor_graph, factor_mean_field)
+
+        self._factor_rescale = factor_rescale
+        self._factor_subset_factor = factor_subset_factor
+        self._ep_mean_field = ep_mean_field
+        self._plates_index = plates_index
+
+    
+    @property 
+    def parameters(self):
+        return {
+            "factor_graph": self._factor_graph, 
+            "factor_mean_field": self._factor_mean_field,
+            "factor_rescale": self._factor_rescale,
+            "factor_subset_factor": self._factor_subset_factor,
+            "ep_mean_field": self._ep_mean_field,
+            "plates_index": self._plates_index,
+        }
+
+    @property 
+    def plates_index(self):
+        return self._plates_index.copy()
+
+    @property 
+    def factor_rescale(self):
+        return self._factor_rescale.copy()
+
+    @property 
+    def factor_subset_factor(self):
+        return self._factor_subset_factor.copy()
+
+    def project_mean_field(
+        self,
+        new_dist: MeanField,
+        factor_approx: FactorApproximation,
+        delta: float = 1.0,
+        status: Status = Status(),
+    ) -> Tuple["EPMeanField", Status]:
+
+        factor_mean_field = self.factor_mean_field
+        
+        # We're fitting the full factor_dist, not the subset factor_dist
+        rescale1 = {
+            v: 1 - scale 
+            for v, scale in self._factor_rescale[factor_approx.factor].items()
+        }
+        last_factor_dist = factor_mean_field.pop(factor_approx.factor)
+        subset_cavity_dist = last_factor_dist.rescale(rescale1)
+        
+        new_factor_dist, status = new_dist.update_factor_mean_field(
+            factor_approx.cavity_dist,
+            factor_approx.factor_dist,
+            delta=delta,
+            status=status,
+        )
+        factor_mean_field[factor_approx.factor] = new_factor_dist  * subset_cavity_dist
+
+        new_approx = type(self)(
+            factor_graph=self._factor_graph,
+            factor_mean_field=factor_mean_field,
+            factor_rescale=self._factor_rescale,
+            factor_subset_factor=self._factor_subset_factor, 
+            ep_mean_field=self._ep_mean_field,
+            plates_index=self._plates_index,
+        )
+        return new_approx, status
+
+    def factor_approximation(self, factor: Factor) -> FactorApproximation:
```

Comment:
Am I confused or is something similar computed elsewhere?

Reply:
Yes, but not for EPMeanFieldSubset objects, so the calculation is subtly different.